### DISCUSSION TITLE
Caching

### DISCUSSION THREAD

**Discussion by duynguyenod** [OngDev/community#237] [Technical Discussion]
Chúng ta có các loại read/write cache bao gồm:
Read: Cache aside, read through.
Write: Write around, write back, write through.

Câu hỏi: cái nào tốt nhất? trường hợp nào nên dùng cái nào?

**Comment by nguyenhuuphu1999** [comment]
anh ơi, cho em hỏi ở phút thứ [15:15](https://www.youtube.com/watch?v=msqPp2J3c6Y&t=915s) anh có nói về chỗ check filter có keys hay không nếu có key mới xuống cache để lấy dữ liệu á anh, thì cái filter đó được đặt ở đâu vậy anh, có phải trong cache luôn không ạ

**Reply by duynguyenod** [reply]
Filter đó đc đặt ở trong app e.

**Reply by dnam-uet** [reply]
vậy tức cái filter đó có thể coi là in memory cache đúng không ạ? Vừa lưu trong app mà không phải biến temp ạ

**Comment by duynguyenod** [comment]
Khum ai trả lời à :v ko biết thì đoán thử xem, tự đưa ra lập luận, rồi anh em trao đổi với nhau?

**Comment by NguyeNgocPhus** [comment]
e lưu vào bộ nhớ cache một số dữ liệu  và đọc dữ liệu từ cache nếu nó tồn tại, nếu không tồn tại thì đọc dữ liệu từ db và ghi dữ liệu đó vào cache. trong trường hợp dữ liệu trong db được update thì dữ liệu trong cache sẽ sử lý ntn ạ. e có tìm hiểu thì có 1 số cách như cache sẽ được update ngay khi db update, hay là xét expired cho dữ liệu trong cache. theo a thì cách nào được dùng trong tình uống này ạ

**Reply by duynguyenod** [reply]
Trong video có phần write đó e, bạn ở dưới cũng giải thích thêm về việc khi nào nên dùng cách nào rồi á

**Comment by cuongnt398** [comment]
- Em nghĩ về phần Cache Aside bị tấn công. Mình không nên set value null vào cache. 
Lý do: Nếu hack nó không spam chỉ 1 ông A, mà nó spam random các ông thì sao ?
-> Vẫn phải gọi vào database liên tục
-> Số lượng key trong cache tăng lên có thể dẫn tới Cache Replacement ( Eviction Policy ) -> Lúc này nghiễm nhiên cả những key đáng lẽ ra có thể sử dụng cũng bị remove đi.
=> Nên sử dụng rate-limiter, hoặc detect ra lỗi đó sớm nhờ việc check số lượng request bị lỗi.

- Về câu hỏi: Chắc chắn là không có 1 phương pháp nào tốt nhất cả. Nó là sự phù hợp, phù hợp ở bài toán và đặc điểm của từng loại cache.

**Cache Aside:** Dữ liệu trong cache và database rõ ràng là không nhất quán. Do khi update dữ liệu thì nó sẽ chỉ update vào database. Để tránh điều này thì data nằm trên cache sẽ phải set TTL ( Time to live). Và sau khi hết hạn thì nó sẽ bị xóa khỏi cache. 
=> Phù hợp cho những hệ thống đọc nhiều. Dữ liệu không cập nhật quá thường xuyên ( như tên, tên tài khoản, email,...) hay dữ liệu cập nhật có thể không quá quan trọng. Và mình có thể set TTL dài cho những data kiểu này.

**Read-Through**: Tương tự Cache Aside, chỉ khác là nói sẽ tự động gọi vào database, cũng như cập nhật vào cache. Thay vì mình phải implement logic chủ động 

**Write-through**: Mỗi khi write vào database thì sẽ phải write vào cache trước. 
Cách này dữ liệu trong cache và database luôn nhất quán, đánh đổi ở đây là tăng độ trễ mỗi khi write vì nó phải write vào cả cache.
=> Phù hợp với những use cases mà data yêu cầu tính nhất quán giữa cache và database. 

**Write-back**: Cách này thì muốn add vào cache, sau đó 1 thời gian mới đồng bộ vào database.
=> Áp dụng để giảm chi phí tránh write vào db quá nhiều. Nói chung cách này thì rủi ro là khi cache die trước khi update vào database thì lost dữ liệu. Cần áp dụng thêm các kĩ thuật khác để tránh rơi vào case trên.

**Reply by duynguyenod** [reply]
@cuongnt398 
A đưa ra một trường hợp để giải thích về cái này nha:
`Em nghĩ về phần Cache Aside bị tấn công. Mình không nên set value null vào cache.
Lý do: Nếu hack nó không spam chỉ 1 ông A, mà nó spam random các ông thì sao ?
-> Vẫn phải gọi vào database liên tục
-> Số lượng key trong cache tăng lên có thể dẫn tới Cache Replacement ( Eviction Policy ) -> Lúc này nghiễm nhiên cả những key đáng lẽ ra có thể sử dụng cũng bị remove đi.
=> Nên sử dụng rate-limiter, hoặc detect ra lỗi đó sớm nhờ việc check số lượng request bị lỗi.`

- Đúng là hướng tiếp cận rate limiter sẽ tốt hơn, với trường hợp em mô tả ở trên và cũng là hầu hết các trường hợp trong thực tế.

Nếu trường hợp chúng ta subscribe tới 1 cái websocket, sử dụng dữ liệu nào đó (trong cache) để xử lý data từ socket, vậy thì lúc này có thể không cần phải là bị tấn công, nhưng nó vẫn vào trường hợp chúng ta nói tới đúng ko?

Tương tự với cronjob.

Em thấy sao?

**Reply by cuongnt398** [reply]
@duynguyenod 
Em nghĩ với trường hợp không bị tấn công thì đúng là mình vẫn set giá trị vào cache. Nhưng thay vì mình set null, mình sẽ set empty value.
Lý do: Thường mình sẽ code như sau.
```
    public static final int DEFAULT_TTL = 5 * 60;
    public Item getItemById(String id) {
        Item item = itemCache.get(id); 
        if(item == null) {
            item = itemRepository.findItemById(id); // null if item isn't exists
            // expire after 5 minutes
            itemCache.put(id, item, DEFAULT_TTL);
        }
        return item;
    }
```
Nếu làm như trên thì đương nhiên với 1 id không tồn tại trong hệ thống, mọi request gửi đế vẫn gọi hết vào database.
=> Sử dụng empty data.
```
    public static final int DEFAULT_TTL = 5 * 60;
    public Item getItemById(String id) {
        Item item = itemCache.get(id);
        if (item == null) {
            item = itemService
                    .findItemById(id)
                    .orElse(Item.empty());

            // expire after 5 minutes
            itemCache.put(id, item, DEFAULT_TTL);
        }
        return item;
    }
```

**Reply by duynguyenod** [reply]
Uh Empty, hoặc là 1 cái value cố định gì đó vẫn là cách tối ưu hơn.

Cái này chỉ là đưa thêm thông tin:
Nếu em đọc code của node cache manager thì:
```
export type Cache<S extends Store = Store> = {
  set: (key: string, value: unknown, ttl?: Ttl) => Promise<void>;
  get: <T>(key: string) => Promise<T | undefined>;
  del: (key: string) => Promise<void>;
  reset: () => Promise<void>;
  wrap<T>(key: string, fn: () => Promise<T>, ttl?: Ttl): Promise<T>;
  store: S;
};
```
thì get sẽ trả về T | undefined, thì thực ra vẫn dùng null đc thôi :v nếu anh dùng ts và compare item === undefined là ok :D

**Comment by solvntech** [comment]
@duynguyenod e đã coi qua 3 cách write cache mà a giới thiệu ở trên video rùi, mà có 1 trường hợp e không pk xử lý thế nào cho nó phù hợp, đó là thời điểm nào mình update lại cache cho nó tối ưu nhất.
E cũng đang suy nghĩ ra 2 cái trường hợp là:
- cách thứ 1: settimeout để update lại cache (bug của nó là dễ outdate khi có rất nhiều required)
- cách thứ 2: subscribe db changed (nếu db thay đổi liên tục thì toang)
Mong ae mình có thể discuss và giúp e có thể mở mang kiến thức đc ko ạ. cảm ơn a rất nhiều ạ.

**Reply by danh-luong** [reply]
Mình có suggest thế này, bạn nên đánh giá về business nữa - liệu stale data có được hiển thị cho user hay ko? Hay có thể chấp nhận stale data trong 1 khoảng thời gian rồi update data mới. 
- cache-aside:  đọc dữ liệu lên set TTL (time quá dài thì stale data - ngắn thì tốn round trip, node fail thì tốn đống thời gian để round trip lại mấy bước trên)
- write-through: cứ có update thì đợi cache update xuống db (thành công - hoặc fail) trả kết quả về cho user (chậm - nhưng consistent data, node fail thì đợi cache retry tốn thêm thời gian)
- write-behind: có update đưa vào cache, cache async đưa xuống db (tốc độ nhanh - nhưng node fail thì inconsistent data, implement phức tạp)
- refresh-ahead: cách này thì predict thôi (có thể dựa trên cache replacement) những entry nào mà đc dùng gần nhất thì sync với db sau 1 khoảng thời gian cố định (set time refresh ngắn thì nặng cache - dài thì stale data)
=> Summary: như những cách trên tuỳ thuộc vào business rule nữa thì mới quyết định phương án tốt nhất được, với cả bạn để ý những thằng write dễ dính vụ lưu quá nhiều data không cần thiết (nên sử dụng TTL - mà TTL thì lại sinh ra vấn đề cache invalidation)

**Comment by TrungAug** [comment]
Em mới tìm hiểu về lập trình, mong anh có thể  clear giúp em giữa session với cache với ạ.
